curator-recipes/src/main/java/org/apache/curator/framework/recipes/shared/SharedValue.java

```diff
@@ -196,8 +196,12 @@ public boolean trySetValue(VersionedValue<byte[]> previous, byte[] newValue) thr
     private void updateValue(int version, byte[] bytes) {
         while (true) {
             VersionedValue<byte[]> current = currentValue.get();
-            if (current.getVersion() >= version) {
-                // A newer version was concurrently set.
```

Comment:
I am not sure wether `SharedValue` was designed to work with multiple owners, but I saw there is a background watcher to update value and also there is no rule to forbid concurrent usages. So, I assume it should work well in case of concurrency.

Then, let me assume a situation:
1. `current.getVersion` is `Integer.MAX_VALUE`.
2. Thread1 call `trySetValue` and succeed to get overflowed version `Integer.MIN_VALUE`, but the call to `updateValue` is somewhat delayed.
3. Thread2 (assume watcher, which runs in ZooKeeper thread if I am not wrong) call `updateVersion` with version `Integer.MIN_VALUE + 1`. According to the code, this will be ignored.
4. Thread1 call `updateValue` to continue its task with version `Integer.MIN_VALUE`. It succeeds.
5. That is all, assume no changes anymore. I know it may not realistic.

`current` stores dated version while the javadoc says "All clients watching the same path will have the up-to-date value (considering ZK's normal consistency guarantees)".

Reply:
I am not convinced in the overflow case either. @tisonkun @Hexiaoqiao 

For the overflow case,
* +1 to deprecate `VersionedValue#getVersion` so to warn clients about the "ordering assumption" if any about `version`.
* +1 to a viable workaround if any and/or exception.

Reply:
I think we should refactor `updateValue` a bit to compare ordering using `Stat.mzxid`. This way we are not fearing this overflow issue. I am stupid in reviwing without a deep thought, sorry for that. So my finally points are:
* Deprecate `VersionedValue#getVersion` to warn clients about "ordering assumptions" and "overflow behavior". 
* Refactor `updateValue` to order using `Stat.mzxid`.
* Throw exception in case of `-1` `Stat.version` in `trySetValue`. I am positive to [ZOOKEEPER-4743](https://issues.apache.org/jira/browse/ZOOKEEPER-4743).
* Document somehow about "overflow" and exception case in `trySetValue`.

Besides above, should we expose a `VersionedValue#getZxid` for client usage ?

Any thoughts @tisonkun @eolivelli @Hexiaoqiao ?

Reply:
I have pushed new commits to go through above direction. Could you please take a look @Hexiaoqiao @tisonkun @eolivelli ?